libs/sync/operators.go

```diff
@@ -0,0 +1,171 @@
+}
+
+func (d operators) IsEmpty() bool {
+	return len(d.put) == 0 && len(d.delete) == 0
+}
+
+// Compute operations required to make files in WSFS reflect current local files.
+// Takes into account changes since the last sync iteration.
+func computeOperators(target *FilesState, current *FilesState) operators {
+	d := &operators{
+		delete: make([]string, 0),
+		rmdir:  make([]string, 0),
+		mkdir:  make([]string, 0),
+		put:    make([]string, 0),
+	}
+	d.caseFilesRemoved(target, current)
+	d.caseRemoteNameChanged(target, current)
+	d.caseFilesAdded(target, current)
+	d.caseFilesUpdated(target, current)
+	return *d
+}
+
+// Add operators for tracked files that no longer exist.
+func (d *operators) caseFilesRemoved(after *FilesState, before *FilesState) {
+	for localName, remoteName := range before.LocalToRemoteNames {
+		if _, ok := after.LocalToRemoteNames[localName]; !ok {
+			d.delete = append(d.delete, remoteName)
+		}
+	}
+
+	// Remove directories that would no longer contain any files.
+	beforeDirs := MakeDirSet(maps.Keys(before.LocalToRemoteNames))
+	afterDirs := MakeDirSet(maps.Keys(after.LocalToRemoteNames))
+	d.rmdir = append(d.rmdir, beforeDirs.Remove(afterDirs).Slice()...)
+}
+
+// Cleanup previous remote files for files that had their remote targets change. For
+// example this is possible if you convert a normal python script to a notebook.
+func (d *operators) caseRemoteNameChanged(after *FilesState, before *FilesState) {
+	for localName, beforeRemoteName := range before.LocalToRemoteNames {
+		afterRemoteName, ok := after.LocalToRemoteNames[localName]
+		if !ok || afterRemoteName == beforeRemoteName {
+			continue
+		}
+		d.delete = append(d.delete, beforeRemoteName)
+	}
+}
+
+// Add operators for files that were not being tracked before.
+func (d *operators) caseFilesAdded(after *FilesState, before *FilesState) {
+	for localName := range after.LastModifiedTimes {
+		if _, ok := before.LastModifiedTimes[localName]; !ok {
+			d.put = append(d.put, filepath.ToSlash(localName))
+		}
+	}
+
+	// Add directories required for these new files.
+	beforeDirs := MakeDirSet(maps.Keys(before.LocalToRemoteNames))
+	afterDirs := MakeDirSet(maps.Keys(after.LocalToRemoteNames))
+	d.mkdir = append(d.mkdir, afterDirs.Remove(beforeDirs).Slice()...)
+}
+
+// Add operators for files which had their contents updated.
+func (d *operators) caseFilesUpdated(after *FilesState, before *FilesState) {
+	for localName, modTime := range after.LastModifiedTimes {
+		prevModTime, ok := before.LastModifiedTimes[localName]
+		if !ok || !modTime.After(prevModTime) {
+			continue
+		}
+		d.put = append(d.put, filepath.ToSlash(localName))
+	}
+}
+
+// groupedMkdir returns a slice of slices of paths to create.
+// Because the underlying mkdir calls create intermediate directories,
+// we can group them together to reduce the total number of calls.
+// This returns a slice of a slice for parity with [groupedRmdir].
+func (d operators) groupedMkdir() [][]string {
```

Comment:
It took a bit of time for me to parse this but if I understood correctly this method is supposed to find and remove paths from d.mkdir which are subpath of any other path in d.mkdir, correct?

If so, we can make a whole function a bit simpler (but less performant)

```
out := d.mkdir
for _, name := range d.mkdir {
  out := slices.DeleteFunc(out, func(n string) bool {
      return strings.HasPrefix(name, n)   
  })
}

```
WDYT @pietern ?

Reply:
Yes, that is correct. The `mkdir` calls create the full directory structure so we can skip intermediates.

Your proposal is O(N^2) though, the current solution is O(2*N).

Reply:
> Your proposal is O(N^2) though, the current solution is O(2*N).

Yes, exactly, that's what I meant by less performant / optimal. Not sure if this optimisation is worth of reduced readability though

Reply:
Resolving this thread because the optimization is unrelated to this PR